Create ATen tensors via TensorOptions

goldsborough opened the pull request:
This PR is a prototype / WIP for a broad change to the way tensors are created in ATen. The fundamental motivation for this PR is to make tensor creation from C++ more like in Python, which means changing the API for factory functions like `at::ones`, as well as teaching ATen about construction axes beyond the backend and scalar type, meaning especially a `device`. This change essentially lays the ground for how tensor creation will happen in c10, and thus defines a new API, currently bolted on top of existing ATen.

This PR proposes the following:

### Creation of a type encapsulating all construction axes of a tensor

At the moment, this type will contain:

* A scalar type,
* A backend,

And in the near future shall contain (and already does somewhat in this prototype):

* A device ordinal,
* A layout (strided, coo_sparse etc.)
* A `requires_grad` boolean.

in code:

```
struct TensorOptions {
        size_t device = 0;
	Layout layout = kStrided;
	ScalarType dtype = kFloat;
        Backend backend = kCPU;
	bool requires_grad = false;
};
```

Each of these fields have default values, such that a default-constructed `TensorOptions` object fully specifies a new tensor.

Furthermore, this class shall have chaining methods:

```
TensorOptions().device(1).dtype(kInt)
```

and there shall be functions with the names of fields, that return a new `TensorOptions` instance:

```
// dtype is a free function that returns a new TensorOptions
auto options = dtype(kFloat).requires_grad(true); 
```

### Use of `TensorOptions` in all factory methods

The new API for tensor creation shall then be (e.g. for `ones`):

```
Tensor ones(IntList size, TensorOptions options = {}) { }
```

with usage examples such as

```
at::ones({2, 2}) // float cpu tensor
at::ones({2, 2}, kInt); // int cpu tensor
at::ones({2, 2}, type(kInt).requires_grad(true)) // int cpu tensor, which sets requires_grad for variables
```

### New code path

This PR also implements a new code path for factory methods. For our recollection, at the moment, we create tensors via e.g.

```
at::ones(at::CPU(at::kFloat), {3, 3})
```

which dispatches to

```
at::CPU(at::kFloat).ones({3, 3});
```

which itself calls

```
at::native::ones(at::CPU(at::kFloat), {3, 3})
```

which then calls

```
auto tensor = at::CPU(at::kFloat).tensor({3, 3});
tensor.fill_(1);
```

This PR proposes to remove factory methods from `Type`, and skip the dispatch via `Type` altogether for all factory methods except `tensor`, which remains the fundamental source of tensors invoked by all other tensor factories:

```
at::ones({3, 3}, at::type(at::kFloat).device(at::CPU))
```

now directly calls

```
at::native::ones({3, 3}, at::type(at::kFloat).device(at::CPU))
```

which now does

```
// `TensorOptions::type()`  returns the type corresponding to its `backend` and `scalar_type`
auto tensor = tensor_options.type().tensor({3, 3});
tensor.fill_(1)
```

At the moment, I have implemented this for `ones`, and I think nothing broke. There should be no implications for the Python API.

@zdevito @apaszke @gchanan @colesbury @ezyang 

Fixes https://github.com/pytorch/pytorch/issues/6285
Fixes https://github.com/pytorch/pytorch/issues/6286
Fixes https://github.com/pytorch/pytorch/issues/7735

apaszke commented on aten/src/ATen/TensorOptions.h:
```diff
@@ -0,0 +1,60 @@
+
+ private:
+  // shold this be optional?
+  size_t device_ = 0;
```
It should. `cuda` without device ordinal means the device currently set in the CUDA context, not device 0.

ezyang commented:
In PyTorch, there is a context-manager which can be used to change the default device options, e.g., changing everything from float to double. Do you plan to eventually support this? If so, with the design as written here you are committing to the fact that these values get read at DeviceOption construction time (since these are direct things, not optional)

goldsborough commented:
That's a very good point. I think it would be easy to implement with thread local globals and a RAII mechanism to set defaults in a particular scope. I would personally think it's fine that values are fixed at the call site (where you write `at::ones(...)`). Do you foresee a need to defer the point in time when values are fixed?

goldsborough commented:
That's Coming

goldsborough commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -3885,7 +3886,7 @@
     if (self_->isScalar()) {
       // Empty tensor
-      return self_->type().toScalarType(kLong).tensor({0});
+      return at::native::empty({0}, getType(toDense(self_->type().backend()), kLong));
```
Here I ran into a weird bug where a the `._indices` of a `SparseTensor` became `SparseTensor`s themselves. I'm not sure why `self_->type().toScalarType(kLong).tensor({0});` used to get rid of the sparseness (it should return a `CUDALongType` for a `SparseCUDA<X>` type.

ezyang replied:
If self is sparse, this is definitely insufficient to allocate a dense tensor, and if it did do that, it's a bug.

goldsborough commented on tools/autograd/gen_python_functions.py:
```diff
@@ -233,25 +247,6 @@ def create_python_bindings(python_functions, has_self, is_module=False):
-                return arg['name']
-        return None
-
-    def auto_gpu(option, has_device_bind):
```
This has moved to `aten/src/ATen/function_wrapper.py` since `AutoGPU` is now in ATen

goldsborough commented:
Hello @zdevito @colesbury @gchanan @apaszke @ezyang

This PR is now ready, and it is fat. Tests all passed locally (though probably issues will arise in CI), and all major functionality is implemented.

Unfortunately, this PR is a mix of important changes, and codemods to update all uses of factory methods to the new API. Therefore, here is a rough list of changes I have made. Please start by reviewing these files, then maybe glance over all the other files. You may also want to review individual commits if that helps.

__Classes I have created__:
- `at::Device` in `aten/src/ATen/Device.{h,cpp}`
- `at::Layout` in `aten/src/ATen/Layout.{h,cpp}`
- `at::AutoGPU` in `aten/src/ATen/AutoGPU.{h,cpp}`
- `at::TensorOptions` in `aten/src/ATen/TensorOptions.{h,cpp}` and `torch/csrc/TensorOptions.h`
- `at::detail::DynamicCUDAInterface` in `aten/src/ATen/detail/CUDAHooksInterface.h` and `aten/src/ATen/detail/CUDAHooks.cpp`

__Important files I have modified__:
- aten/src/ATen/native/TensorFactories.cpp
- aten/src/ATen/native/native_functions.yaml
- tools/autograd/gen_python_functions.py
- aten/src/ATen/function_wrapper.py
- tools/autograd/gen_variable_type.py
- tools/jit/gen_jit_dispatch.py
- torch/csrc/utils/tensor_new.cpp

__Some notes__:

Factory functions (like `at::ones`) used to take the type as their first argument and then the size second. This is now changed that the size list is always first, and TensorOptions follow. This is like to PyTorch.

To allow default arguments, while also have the `TensorOptions` always be the last argument, I have changed all factory methods that took default arguments to follow this kind of scheme:
1. `func(arg1, arg2, options)`
2. `func(arg1, arg2, default_arg1, options)`
3. `func(arg1, arg2, default_arg1, default_arg2, options)`
etc. This is a small hassle to implement in `TensorFactories.cpp`, but leads to a nice, uniform API.

AutoGPU has now moved entirely into ATen, and is placed inside e.g. `at::ones` before it calls into `at::native::ones`.

I have opted for `at::optional<int32_t>` as the device index type instead of `int32_t`, such that the default device is represented by an empty optional instead of -1. This is more principled and makes for nicer and safer (!) code. To not disturb other codepaths too much, I made AutoGPU nevertheless respect -1 as the default device.

__TODO__:
- A mechanism to default tensor options thread locally, like in PyTorch (e.g. `with torch.cuda.device()`). See `test/cpp/api/tensor.cpp` for notes on this.

goldsborough commented:
Caffe2 has some kind of ATen Op thing, which I have to fix.

dzhulgakov commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,107 @@
+/// device index, if one is explicitly stored:
+/// 1. The device index is never negative.
+/// 2. When the device type is CPU, the device index must be zero.
+struct Device {
```
So from convergence standpoint - this is pretty much DeviceOption from Caffe2. No action needed, just pointing out

dzhulgakov commented on torch/csrc/TensorOptions.h:
```diff
@@ -0,0 +1,110 @@
+#include <ATen/Type.h>
+
+namespace torch {
+struct TensorOptions : public at::TensorOptions {
```
we're going in the direction of folding autograd into core Tensor class. If that's the case - shall requires_grad be present just in the parent class so that we can avoid 2 different types?

dzhulgakov replied:
Though then apply() won't work..

goldsborough replied:
Yes `torch::TensorOptions` exists mainly to solve the problem of setting `requires_grad` for new variables, which I've achieved via polymorphism here. I figured it's better to only allow setting `requires_grad` when using this subclass, then. But it could just as well be in ATen `TensorOptions`

zdevito replied:
My concern with this subclassing is that we are making something that frequently will just be constant-propagated away into a bunch of virtual function calls. Can we make requires_grad an optional<bool>?

true - Variable requires grad
false - Variable no grad
none - tensor

Like at::CPU vs torch::CPU we can make the default based on which function you used to create the TensorOption.

dzhulgakov commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -21,72 +17,89 @@
+#include <stdexcept>
+#include <vector>
+
+// Certain types have the same name in torch:: and at::, so better use
```
this part is actually quite fragile. Why not name classes separately? E.g. it could be VariableTensorOptions instead of torch::TensorOptions

goldsborough replied:
I might want to call it `VariableOptions`, yes

dzhulgakov reviewed (commented):
Looks good. Some minor comments/questions on how the classes would evolve going forward

goldsborough commented on aten/src/ATen/AutoGPU.h:
```diff
@@ -0,0 +1,96 @@
+/// The index is represented by an `optional<int32_t>`. Both `nullopt` and `-1`
+/// represent the default device. `nullopt` should be preferred, support for
+/// `-1` is kept for legacy reasons.
+struct AutoGPU {
```
I think calling it `DeviceGuard` and making it suitable for more general devices makes sense,  yes! I'll implement that

zdevito commented on aten/src/ATen/test/atest.cpp:
```diff
@@ -9,7 +9,7 @@ using namespace std;
 
 void trace() {
-  Tensor foo = rand(CPU(kFloat), {12,12});
+  Tensor foo = rand({12,12});
```
I think it would be worth it to add backwards compatibility functions to TensorOption for
`ones, zeros, rand, tensor`. That way `at::CPU(at::kFloat).ones()` still works, for example. These can be written manually and forward to the `at::ones` versions. If you look at these tests it is nearly 25% of the lines that need modification, and I worry how much we will break our few C++ users. Those 4 functions are one liners and handle the majority of cases.

goldsborough replied:
Good idea

zdevito commented on test/cpp/api/tensor.cpp:
```diff
@@ -0,0 +1,23 @@
+
+TEST_CASE("tensor/device-placement") {
+  SECTION("AutoGPU") {
+    // SECTION("On index zero by default") {
```
Why is this commented out?

zdevito commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -13,8 +13,8 @@ Linear::Linear(size_t features_in, size_t features_out)
   weight_ =
-      register_parameter("weight", at::CPU(at::kFloat).empty({out_, in_}));
-  bias_ = register_parameter("bias", at::CPU(at::kFloat).empty(out_));
+      register_parameter("weight", at::empty({out_, in_}));
```
These changes needed to the C++ API are a good sentinel for how much we will be breaking people, and what core functions we should maintain backward compatibility for.

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -46,16 +46,20 @@
 auto result = at::${name}(${args});
 """)
 CALL_METHOD = CodeTemplate("""\
 AutoGPU device_guard(deviceForInputs(stack, ${num_dynamic_inputs}));
```
Is this still needed or do the ATen ops now correctly set the device on their own?

zdevito commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -429,11 +440,11 @@ void interpTest() {
-    auto w_ih  = t_def(at::randn(at::CUDA(at::kFloat), {4 * hidden_size, input_size}));
-    auto w_hh  = t_def(at::randn(at::CUDA(at::kFloat), {4 * hidden_size, hidden_size}));
+    auto input = at::randn({seq_len, batch_size, input_size}, at::device(at::kCUDA));
+    auto hx    = at::randn({batch_size, hidden_size}, at::device(at::kCUDA));
```
I'd be nice if you could write something shorter than `at::device(at::kCUDA)`

zdevito reviewed (commented):
This looks really great! I have some questions and some suggestions, but they are minor compared to the overall design and implementation.  Once this is in, we can simplify the aten schema, since we no longer have public aten functions that are specified on the type, only either "this is a function" or "this is a method", which would be neat to clean up.

goldsborough commented on aten/src/ATen/AutoGPU.h:
```diff
@@ -0,0 +1,96 @@
+  /// Resets the GPU to the index that was active at construction of the guard.
+  ~AutoGPU() {
+    // It should only not have a value if an index was never actually set.
+    if (original_index_.has_value()) {
```
I guess it's more explicit, but I can change it

goldsborough commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -441,10 +447,10 @@
   variants: function
   python_default_init:
-    dtype: self.type()
+    options: self.type()
```
The thing is that currently `Type` holds the additional piece information about whether to construct a `Variable` vs. a `Tensor`. Also an empty `TensorOptions` would change any variable to `CPU(Float)` by default ...

colesbury replied:
I'm a bit confused by this. I don't see any corresponding `self.type()` in the generated Python binding code for this function. Why does this need any `python_default_init`?

colesbury reviewed (commented):
To parrot @zdevito, I'd would prefer one `TensorOptions` class. I'd be a little concerned about the overhead of chaining all the calls of TensorOption that currently can't be inlined.

I'm a bit concerned about adding AutoGPU calls everywhere in the ATen interface. This means that when ATen functions call other ATen functions there's an AutoGPU call at every layer where previously there would be only one call at the Python binding level. There seems to be a performance regression with indexing: ([loop.py](https://gist.github.com/colesbury/99e3bffd89a72adb6af6ac132d89e715)).

In terms of `DynamicCUDAInterfaceSetter`, the current design is fine, but you can probably avoid the awkward reinterpret casts by defining wrappers in CUDAHooks.cpp that also check the error code. Something like:

```
void DynamicCUDAInterface::set_device(int device) {
  check_status(cudaSetDevice(device)); 
}
int DynamicCUDAInterface::get_device() { ... }
void DynamicCUDAInterface::set_device_unchecked(int device) { ... }
``` 

I know this is different from the advice I gave originally -- I probably gave bad advice. The number of calls that can't be inlined doesn't change much since check_status is pushed down, and regardless, `cudaGetDevice` is more expensive (~35-40ns) than I remembered compared to the wrapper call. The most important thing is that there is negligible overhead for non-CUDA tensors which is still the case.

colesbury commented on torch/csrc/utils/tensor_new.h:
```diff
@@ -1,14 +1,20 @@
 
 #include "torch/csrc/python_headers.h"
-#include "torch/csrc/utils/device.h"
-#include <ATen/ATen.h>
```
Is there a reason to remove the header? You can't do anything with `tensor_new.h` without the full Tensor struct definition (i.e. <ATen/ATen.h>)

goldsborough commented:
The main purpose of `TensorOptions` is to return a `VariableType` inside `TensorOptions::type`. How else can I create a `Variable` when calling an ATen function, without wrapping the function call in a variable construction? I.e. if all I have is `at::ones(...)` how can I create a `Variable` from within ATen? The only way I saw was this way. @colesbury

colesbury commented:
@goldsborough, I don't a better idea other than a global hook sort of like DynamicCUDAInterface. This isn't great, but it mostly remains an implementation detail that I imagine will get better with C10. I'm curious about @zdevito's opinion.

As a minor note, there are currently two TensorOption functions that depend on Variable behavior: type() and apply(). You can just have one: `Tensor wrap(Tensor)`. The `type()` function could always return the non-Variable Tensor type and `Tensor wrap(Tensor)` would wrap it in a Variable and set requires_grad if necessary. This isn't too important, it just avoids the `set_version_counter(0)`.

I think a single `TensorOption` struct will segue more easily into C10 as well, which won't have the same stratification between Variable/Tensor. (cc @ezyang, correct me if I'm wrong).

goldsborough committed:
Created TensorOptions

Storing the type in TensorOptions to solve the Variable problem

Created convenience creation functions for TensorOptions and added tests

Converted zeros to TensorOptions

Converted rand to TensorOptions

Fix codegen for TensorOptions and multiple arguments

Put TensorOptions convenience functions into torch namespace too

All factory functions except *_like support TensorOptions

Integrated with recent JIT changes

Support *_like functions

Fix in place modification

Some cleanups and fixes

Support sparse_coo_tensor

Fix bug in Type.cpp

Fix .empty calls in C++ API

Fix bug in Type.cpp

Trying to fix device placement

Make AutoGPU CPU compatible

Remove some auto_gpu.h uses

Fixing some headers

Fix some remaining CUDA/AutoGPU issues

Fix some AutoGPU uses

Fixes to dispatch_tensor_conversion

Reset version of new variables to zero

Implemented parsing device strings

Random fixes to tests

Self review cleanups

flake8

Undo changes to variable.{h,cpp} because they fail on gcc7.2

Add [cuda] tag to tensor_options_cuda.cpp

Move AutoGPU::set_index_from into .cpp file because Windows is stupid and sucks

Fix linker error in AutoGPU.cpp

Fix bad merge conflict in native_functions.yaml

Fixed caffe2/contrib/aten

Fix new window functions added to TensorFactories.cpp

goldsborough commented:
Hello again. I've made lots of updates after everyone's useful comments. Here is a list:

1. Removed `torch::TensorOptions` and instead add code to generated `torch::{ones, zeros, ...}` that does the variable wrapping and setting of `requires_grad` inside. In the future these methods can just be aliases for those in c10. At the moment, it solves the `Variable`/`Tensor` problem without incurring any virtual dispatch overhead. @zdevito @colesbury 
2. Generalized `AutoGPU` lightly into `DeviceGuard` and codemodded everything to use this class instead. This was in response to @dzhulgakov's comment and also makes sense for C++ users since we can transparently give `DeviceGuard` more functionality as more device types become available.
3. Added back all the old factory methods, but with deprecation attributes which give nice warnings when they are used. @zdevito @gchanan

goldsborough commented:
@pytorchbot retest this please

goldsborough committed:
Removed torch::TensorOptions

Added code to generate wrapper functions for factory methods

Add implicit constructor from Backend to TensorOptions

Remove Var() from C++ API and use torch:: functions

Use torch:: functions more subtly in C++ API

Make AutoGPU::set_device more exception safe

Check status directly in DynamicCUDAHooksInterface

Rename AutoGPU to DeviceGuard

Removed set_requires_grad from python_variables.h and warn appropriately in Variable::set_requires_grad

remove python_default_init: self.type()

Add back original factory functions, but with deprecation warnings

Disable DeviceGuard for a couple functions in ATen

Remove print statement

Fix DeviceGuard construction from undefined tensor

Fixing CUDA device compiler issues

Moved as many methods as possible into header files

Dont generate python functions for deprecated factories

Remove merge conflict artefact

Fix tensor_options_cuda.cpp

Fix set_requires_grad not being checked

Fix tensor_new.h

TEMPORARILY put some methods in .cpp files to see if it solves issues on windows and mac

Fix bug in DeviceGuard.h

Missing includes

TEMPORARILY moving a few more methods into .cpp to see if it fixes windows

Fixing linker errors

colesbury commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,118 @@
+
+  /// Constructs a new `Device` from a `Tensor`'s type and, if it is a CUDA
+  /// tensor, its device index.
+  explicit Device(Tensor tensor);
```
`const Tensor& tensor`?

colesbury commented on aten/src/ATen/DeviceGuard.h:
```diff
@@ -0,0 +1,111 @@
+        detail::DynamicCUDAInterface::get_device(&previous_index);
+        original_device_ = Device(Device::Type::CUDA, previous_index);
+      } else {
+        original_device_ = Device(Device::Type::CPU);
```
This looks wrong to me. Consider this sequence of calls:

```
DeviceGuard guard;
guard.set_device(/*CPU*/);
guard.set_device(1);
```

When the DeviceGuard is destroyed, it won't restore the correct CUDA device because original_device will be a CPU device.

gchanan commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,113 @@
+/// "the current device". Further, there are two constraints on the value of the
+/// device index, if one is explicitly stored:
+/// 1. The device index is never negative.
+/// 2. When the device type is CPU, the device index must be zero.
```
should this be..."when the device type is CPU and a device ordinal exists,it must be 0?

gchanan commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,113 @@
+  /// defaults to `nullopt`. If no device index is given, the device
+  /// type defaults to CPU. If a device index is supplied, the type defaults to
+  /// CUDA, such that a string like "1" would be equivalent to "cuda:1".
+  /* implicit */ Device(const std::string& device_string);
```
do we need this in the API?  I put something similar in the pytorch API for backwards compatibility reasons, but we don't really have devices in ATen yet, so that doesn't seem to be a concern.  Can we just bridge where we need it in pytorch?

goldsborough replied:
I thought it would be nice for research users. It's not connected to the Python implementation

goldsborough replied:
Oh I see you mean the behavior with only a digit

goldsborough replied:
Good point, I should remove that

gchanan replied:
right, sorry for not being clear.  I expected just an integer to be self contained at the PyTorch level.

gchanan commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,113 @@
+
+  /// Constructs a new `Device` from a `Backend` (which is converted to a
+  /// `Type`, if possible) and an optional device index.
+  /* implicit */ Device(Backend backend, at::optional<int32_t> index);
```
why have any of these be implicit?

goldsborough replied:
So you can write `at::device(kCUDA).dtype(...)`

gchanan commented on aten/src/ATen/Device.h:
```diff
@@ -0,0 +1,113 @@
+  /* implicit */ Device(Backend backend);
+
+  /// Constructs a new `Device` from a `Tensor`'s type and, if it is a CUDA
+  /// tensor, its device index.
```
why isn't this just a method on tensor?  `device()` or something?

gchanan commented on aten/src/ATen/DeviceGuard.cpp:
```diff
@@ -0,0 +1,41 @@
+  set_index(index);
+}
+
+/// Legacy constructor that accepts -1 as the device index and turns it into
```
why are we putting legacy stuff in ATen instead of in PyTorch?

gchanan commented on aten/src/ATen/DeviceGuard.h:
```diff
@@ -0,0 +1,85 @@
+/// RAII guard that sets a certain default GPU index in its constructor, and
+/// changes it back to the device that was originally active upon destruction.
+///
+/// The index is always reset to the one that was active at the time of
```
Is this standard with RAII guards?  I wouldn't know if the reset behavior was the previous or first device, which means I have to read the documentation carefully.  But maybe this is standard and I don't know.

Also, if it's standard, is it particularly useful?  Can you point to an example?

goldsborough replied:
This was the previous behavior of `AutoGPU`. I also think that instantiation of a `DeviceGuard` should be like an anchor/checkpoint, and whatever you do you get back to this checkpoint.

gchanan replied:
That sounds fine, it just always seemed a little strange to me that we allow `set_index` instead of just using another guard.

gchanan commented on aten/src/ATen/DeviceGuard.h:
```diff
@@ -0,0 +1,85 @@
+
+  /// Sets the device to the index on which the first tensor in the list is
+  /// located. If the list is empty, does nothing.
+  explicit DeviceGuard(const TensorList& tensors);
```
can we just do this explicitly?  IMO it's better to make it obvious what's happening at the call site than not.

goldsborough replied:
I think I looked into this originally, but the issue is this constructor is used in generated code in `python_torch_functions_dispatch.h`, and it's much more convenient to just use the same code for all dispatch functions instead of changing the codegen

gchanan commented on aten/src/ATen/TensorOptions.h:
```diff
@@ -0,0 +1,186 @@
+
+  /// Constructs the `TensorOptions` from a type and optional `device_index`.
+  ///
+  /// If `discard_runtime_type` is false (the default), the behavior of
```
I'm okay with this, but I wonder if there's a cleaner way.  You really need the type only to know if it's a variable (and because even if you knew this you can't recover it?).  But whether it's a variable or not is already part of the public ATen API.  We can't take something that allows us to do this while still retaining the API we "want" -- layout, device, requires_grad (why is requires_grad not here)?

zdevito replied:
Would `std::optional<bool> requires_grad` work? If it is not present, it is a Tensor, otherwise it is a variable with the equivalent requires_grad setting.

gchanan replied:
Sounds reasonable but @goldsborough mentioned elsewhere he had trouble with optional and NVCC.

goldsborough commented:
Ok, so I've surrendered and undone my changes to make `at::optional<int32_t>` the device index type. The CUDA device compiler is not able to compile it, and I've wasted way too much time trying to move certain methods into the `.cpp` file and fighting these issues on Windows, where they are especially prevalent. -1 is once more default device.

ezyang commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -0,0 +1,44 @@
+        if argument["simple_type"] == "TensorOptions":
+            # We want to discard the runtime type so that `at::{name}` always returns a
+            # tensor and not a variable, since we create a variable right after.
+            actual += ".discard_runtime_type()"
```
I'm skeptical. Why are you "creating a variable right after"?

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -13,22 +15,61 @@
+// Single, global, static (because of the anonymous namespace) instance, whose
+// constructor will set the static members of `DynamicCUDAInterface` to CUDA
+// functions when the ATen CUDA library is loaded.
+DynamicCUDAInterfaceSetter _;
```
They do.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -142,7 +171,7 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 ZERO_DIM_ONLY = CodeTemplate("""\
 AT_ERROR("${api_name} only supports a 0-dimensional ${check_name} tensor, but got tensor "
-    "with %" PRId64 " dimension(s)", ${check_name}.dim());
+    "with ", ${check_name}.dim(), " dimension(s).");
```
Thanks

ezyang reviewed (approved):
ship it already yo

zdevito reviewed (approved):
There are some details about TensorOptions like the underlying Type* field that we might want to figure out better solutions for. However, this PR is getting to big, so assuming there are not any known bugs, I think we should merge and revisit those nits afterwards.

goldsborough committed:
Fix up SummaryOps to use new factories

Undo device agnostic behavior of DeviceGuard

Use -1 instead of optional for default device index

Also move DeviceGuard methods into header

Fixes around device index after optional -> int32_t switch

Fix use of DeviceGuard in new_with_tensor_copy

Fix tensor_options.cpp

goldsborough committed:
Fix Type::copy(

goldsborough committed:
Remove test_non_float_params from ONNX tests

goldsborough committed:
Set requires_grad=False in ONNX tests that use ints

goldsborough committed:
Put layout/dtype/device on Tensor

goldsborough committed:
Post merge fixes

goldsborough committed:
Change behavior of DeviceGuard to match AutoGPU

goldsborough committed:
Fix C++ API integration tests

goldsborough committed:
Fix flip functions

colesbury commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -577,7 +576,7 @@ void DataChannelNccl::broadcast(std::vector<at::Tensor>& data,
 
   // Guard GPU device
-  AutoGPU gpuGuard;
+  at::DeviceGuard gpuGuard;
```
@goldsborough this does not compile. See line 587

goldsborough replied:
fixing

SsnL commented on aten/src/ATen/TensorOptions.h:
```diff
@@ -0,0 +1,208 @@
+    } else {
+      backend = (layout_ == kStrided) ? kCUDA : kSparseCUDA;
+    }
+    return getType(backend, dtype_);
```
I see. Thanks for explaining!
